lib/libm/man/log.3
.\" $NetBSD: log.3,v 1.8 2024/01/26 19:27:30 nros Exp $
.\"
.\" Copyright (c) 2011 Jukka Ruohonen <jruohonen@iki.fi>
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
.\" ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
.\" TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
.\" PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
.\" BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
.\" CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\" ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd January 24, 2024
.Dt LOG 3
.Os
.Sh NAME
.Nm log ,
.Nm logf ,
.Nm logl ,
.Nm log10 ,
.Nm log10f ,
.Nm log10l ,
.Nm log1p ,
.Nm log1pf ,
.Nm log1pl ,
.Nm log2 ,
.Nm log2f
.Nm log2l
.Nd logarithm functions
.Sh LIBRARY
.Lb libm
.Sh SYNOPSIS
.In math.h
.Ft double
.Fn log "double x"
.Ft float
.Fn logf "float x"
.Ft long double
.Fn logl "long double x"
.Ft double
.Fn log10 "double x"
.Ft float
.Fn log10f "float x"
.Ft log double
.Fn log10l "long double x"
.Ft double
.Fn log1p "double x"
.Ft float
.Fn log1pf "float x"
.Ft long double
.Fn log1pl "long double x"
.Ft double
.Fn log2 "double x"
.Ft float
.Fn log2f "float x"
.Ft long double
.Fn log2l "long double x"
.Sh DESCRIPTION
The following functions compute logarithms:
.Bl -bullet -offset 2n
.It
The
.Fn log ,
.Fn logf
and
.Fn logl
functions return the natural logarithm.
.It
The
.Fn log10 ,
.Fn log10f
and
.Fn log10l
functions return the base 10 logarithm.
.It
The
.Fn log1p ,
.Fn log1pf
and
.Fn log1pl
functions return the natural logarithm of (1.0 +
.Fa x )
accurately even for very small values of
.Fa x .
.It
The
.Fn log2 ,
.Fn log2f
and
.Fn log2l
functions return the base 2 logarithm.
.El
.Sh RETURN VALUES
Upon successful completion, the functions return the logarithm of
.Fa x
as described above.
Otherwise the following may occur:
.Bl -enum -offset indent
.It
If
.Fa x
is \*(Na, all functions return \*(Na.
.It
If
.Fa x
is positive infinity, all functions return
.Fa x .
If
.Fa x
is negative infinity, all functions return \*(Na.
.It
If
.Fa x
is +0.0 or -0.0, the
.Fn log ,
.Fn log10 ,
and
.Fn log2
families return either
.Dv -HUGE_VAL ,
.Dv -HUGE_VALF ,
or
.Dv -HUGE_VALL ,
whereas the
.Fn log1p
family returns
.Fa x .
.It
If
.Fa x
is +1.0, the
.Fn log ,
.Fn log10 ,
and
.Fn log2
families return +0.0.
If
.Fa x
is -1.0, the
.Fn log1p
family returns
.Dv -HUGE_VAL ,
.Dv -HUGE_VALF ,
or
.Dv -HUGE_VALL .
.El
.Pp
In addition, on a
.Tn VAX ,
.Va errno
is set to
.Er EDOM
and the reserved operand is returned
by
.Fn log
unless
.Fa x
> 0, by
.Fn log1p
unless
.Fa x
> \-1.
.Sh SEE ALSO
.Xr exp 3 ,
.Xr ilogb 3 ,
.Xr math 3
.Sh STANDARDS
The described functions conform to
.St -isoC-99 .
.Sh HISTORY
A
.Fn log
function appeared in
.At v1 .


doc/CHANGES
# LIST OF CHANGES FROM LAST RELEASE:			<$Revision: 1.3032 $>
#
#
# [Note: This file does not mention every change made to the NetBSD source tree.
# For an exhaustive list, please subscribe to the `source-changes' mailing list,
# or see the mailing list archives.  For more information, send a message
# containing just the word `help' to `majordomo@NetBSD.org'.]
#
# ----------------------------------------------------------------------------
# PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:
#
#	theme: Content ...
#		more content ...
#		more content. [developer DATE-FORMAT]
#
# That's it, one tab to start an item, and two tabs to indent the content.
# Also please start the content with a capital and end it with a dot.
#
# This is very important for the www scripts, otherwise it won't be able to
# parse the file and the HTML output will be mangled.
#
# See htutils/changes/changes2html script for more details.
#
# ----------------------------------------------------------------------------

Changes from NetBSD 10.0 to NetBSD 11.0:
	amdsmn(4): added trivial power management support [reinoud 20221218]
	amdccp(4): added trivial power management support [reinoud 20221218]
	ppp(4): Remove several non-functioning ioctls. [chs 20221221]
	dse(4): Added Dayna SCSI/Link ethernet driver. [nat 20221222]
	binutils: Updated to FSF binutils 2.39. [christos 20221224]
	cribbage(6): Add -y option to keep playing without asking
		(from Greywolf in PR 57164) [christos 20230106]
	x86: Enable the -z separate-code security feature by default in
		ld(1). Reduces the default max page size from 2M to
		4K on amd64. [christos 20230106]
	bind: Import version 9.16.37. [christos 20230125]
	xfwp(1): remove. [wiz 20230204]
	wscons(4): Ignore nonsense tab stops in vt100 emulation.
		[riastradh 20230223]
	nvmm(4): Filter CR4 bits on x86 SVM (AMD). In particular,
		prohibit PKE, Protection Key Enable, which requires
		some additional management of CPU state by nvmm. 
		[riastradh 20230223]
	aarch64: Audit use of curcpu(9). [riastradh 20230225]
	xen: Memory barrier audit. Reduce unnecessary membars.
		[riastradh 20230225]
	x86: Mitigate MXCSR Configuration Dependent Timing CPU security
		flaw in kernel FPU use. [riastradh 20230225]
	kernel: Optimization: Omit needless membar when triggering softint
		in various ports. [riastradh 20230301]
	sh(1): Adjust tilde expansion to comply with upcoming POSIX
		Issue 8 requirements. [kre 20230306]
	libm: Add more long double transcendental functions (as aliases of
		double functions) from C99. [riastradh 20230313]
	xen: Allow a PVH dom0 to use VGA as console. [bouyer 20230324]
	vioscsi(4), viornd(4), viomb(4): Mark VirtIO drivers as MPSAFE.
		[mlelstv 20230325]
	amiga: Fix bootblocks for Kickstart 3.2. [abs 20230325]
	libc: Add __atomic_is_lock_free to help stdatomic.h usage with
		GCC. [riastradh 20230330]
	vacation(1): Check 'Auto-Submitted:' header in addition to
		'Precedence:', and set 'Precedence:' in addition to
		'Auto-Submitted:'.  [hauke 20230411]
	ichsmb(4), tco(4): Add support for TCO on newer Intel chipsets.
		[riastradh 20230412]
	audiorecord(1): Add support for recording 24-bit wav files.
		[mlelstv 20230415]
	evbarm: Install a boot.cfg file for the EFI bootloader by default.
		[brook 20230419]
	dhcpcd: Import version 10.0.1. [roy 20230421]
	evbarm: Enable rkv1crypto driver on Rock64. [mrg 20230424]
	worms(6): Add a little bit of optional colour, code and
		ideas from RVP. [kre 20230426]
	amiga: Add driver for the MNT ZZ9000 Zorro board - submitted by
		Alain Runa. [phx 20230503]
	evbarm: Experimental IMX6SX support - tested on the UDOO Neo Full board.
		[bouyer 20230504]
	chown(8): Add a -d flag to avoid changing a file's owner/group to the
		current value.  This avoids some unnecessary operations on the
		file. [pgoyette 20230504]
	riscv: Add RISC-V support that works on QEMU with a single hart.
		[skrll 20230507]
	riscv: Add Google Goldfish RTC driver. [skrll 20230508]
	make(1): Make :mtime operate on each word in variable
		value. [sjg 20230510]
	worms(6): When using -H, make the appearance of the worm head
		depend upon the body (more than it did before). [kre 20230512]
	amiga: Update loadbsd source and distribution binary to version 3.3.
		[phx 20230514]
	i386: Add /dev/efi device. [riastradh 20230522]
	efi(4): Add EFIIOC_GET_TABLE ioctl support on x86. [riastradh 20230522]
	ctfmerge(1): Add error handling to pthread APIs.
		Fix semaphores for macOS host tool.  [lukem 20230523]
	share/mk: Deprecate non-functional SHAREDSTRINGS. [lukem 20230524]
	mkstr(1), xstr(1): Obsolete unused PDP-11-era tools. [lukem 20230524]
	pthread(3): Reduce power consumption when spinning for a lock on
		arm. [riastradh 20230525]
	pax(1): Exit 0 if stdin filelist is used and no files are supplied.
		Don't overwrite existing destination files if -r -w copy
		fails for any reason. [lukem 20230528]
	date(1): Add -R option for displaying time in RFC 5322 format,
		similar to GNU date. [kim 20230531]
	mk.conf(5): Obsolete non-MK variants MANZ UNPRIVED UPDATE
		[lukem 20230602]
	build.sh: Add show-params operation [lukem 20230602]
	kerberos(8): Update to Heimdal-7.8.0 [christos 20230619]
	patch(1): Add --backup-if-mismatch and --no-backup-if-mismatch for
		GNU patch compatibility. These options only make sense in
		POSIX mode, patch(1) has --backup enabled by default and
		GNU patch doesn't. [wiz 20230616]
	indent(1): Remove support for handling ASCII backspace characters
		in C code. [rillig 20230618]
	make(1): Recognize include guards. [rillig 20230619]
	editline(3): Install pkg-config files, same as the portable version.
		[wiz 20230620]
	make(1): Speed up pattern matching in the ':M' and ':N' modifiers.
		[rillig 20230622]
	paxctl(8): Introduce -0 option to clear all PaX flag bits in ELF note.
		[rin 20230623]
	crunchgen(1): Clear PaX flags instead of removing its ELF note section.
		[rin 20230623]
	make(1): Warn about malformed patterns in ':M', ':N' and

games/atc/log.c
/*	$NetBSD: log.c,v 1.25 2021/05/02 12:50:43 rillig Exp $	*/

/*-
 * Copyright (c) 1990, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * Ed James.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

/*
 * Copyright (c) 1987 by Ed James, UC Berkeley.  All rights reserved.
 *
 * Copy permission is hereby granted provided that this notice is
 * retained on all partial or complete copies.
 *
 * For more info on this and all of my stuff, mail edjames@berkeley.edu.
 */

#include <sys/cdefs.h>
#ifndef lint
#if 0
static char sccsid[] = "@(#)log.c	8.1 (Berkeley) 5/31/93";
#else
__RCSID("$NetBSD: log.c,v 1.25 2021/05/02 12:50:43 rillig Exp $");
#endif
#endif /* not lint */

#include <sys/types.h>
#include <sys/utsname.h>
#include <sys/stat.h>	/* for umask(2) */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <pwd.h>
#include <err.h>

#include "pathnames.h"
#include "def.h"
#include "struct.h"
#include "extern.h"
#include "tunable.h"

static FILE *score_fp;

static int
compar(const void *va, const void *vb)
{
	const SCORE	*a, *b;

	a = (const SCORE *)va;
	b = (const SCORE *)vb;
	if (b->planes == a->planes)
		return (b->time - a->time);
	else
		return (b->planes - a->planes);
}

#define SECAMIN		60
#define MINAHOUR	60
#define HOURADAY	24
#define SECAHOUR	(SECAMIN * MINAHOUR)
#define SECADAY		(SECAHOUR * HOURADAY)
#define DAY(t)		((t) / SECADAY)
#define HOUR(t)		(((t) % SECADAY) / SECAHOUR)
#define MIN(t)		(((t) % SECAHOUR) / SECAMIN)
#define SEC(t)		((t) % SECAMIN)

static const char *
timestr(int t)
{
	static char	s[80];

	if (DAY(t) > 0)
		(void)snprintf(s, sizeof(s), "%dd+%02dhrs", DAY(t), HOUR(t));
	else if (HOUR(t) > 0)
		(void)snprintf(s, sizeof(s), "%d:%02d:%02d", HOUR(t), MIN(t),
			SEC(t));
	else if (MIN(t) > 0)
		(void)snprintf(s, sizeof(s), "%d:%02d", MIN(t), SEC(t));
	else if (SEC(t) > 0)
		(void)snprintf(s, sizeof(s), ":%02d", SEC(t));
	else
		*s = '\0';

	return (s);
}

void
open_score_file(void)
{
	mode_t old_mask;
	int score_fd;
	int flags;

	old_mask = umask(0);
#if defined(O_NOFOLLOW)
	score_fd = open(_PATH_SCORE, O_CREAT|O_RDWR|O_NOFOLLOW, 0664);
#else
	score_fd = open(_PATH_SCORE, O_CREAT|O_RDWR, 0664);
#endif
	(void)umask(old_mask);
	if (score_fd < 0) {
		warn("open %s", _PATH_SCORE);
		return;
	}
	if (score_fd < 3)
		exit(1);
	/* Set the close-on-exec flag.  If this fails for any reason, quit
	 * rather than leave the score file open to tampering.  */
	flags = fcntl(score_fd, F_GETFD);
	if (flags < 0)
		err(1, "fcntl F_GETFD");
	flags |= FD_CLOEXEC;
	if (fcntl(score_fd, F_SETFD, flags) == -1)
		err(1, "fcntl F_SETFD");
	/*
	 * This is done to take advantage of stdio, while still
	 * allowing a O_CREAT during the open(2) of the log file.
	 */
	score_fp = fdopen(score_fd, "r+");
	if (score_fp == NULL) {
		warn("fdopen %s", _PATH_SCORE);
		return;
	}
}

int
log_score(int list_em)
{
	int		i, num_scores = 0, good, changed = 0, found = 0;
	struct passwd	*pw;
	char		*cp;
	SCORE		score[100], thisscore;
	struct utsname	lname;
	long		offset;

	if (safe_planes == 1)
		printf("You directed 1 plane safely to its destination.\n\n");
	else
		printf("You directed %d planes safely to their destinations.\n\n",
		    safe_planes);

	if (score_fp == NULL) {
		warnx("no score file available");
		return (-1);
	}

#ifdef BSD
	if (flock(fileno(score_fp), LOCK_EX) < 0)
#endif
#ifdef SYSV
	if (lockf(fileno(score_fp), F_LOCK, 1) < 0)
#endif
	{
		warn("flock %s", _PATH_SCORE);
		return (-1);
	}
	for (;;) {
		good = fscanf(score_fp, SCORE_SCANF_FMT,
			   score[num_scores].name,
			   score[num_scores].host,
			   score[num_scores].game,
			   &score[num_scores].planes,
			   &score[num_scores].time,
			   &score[num_scores].real_time);
		if (good != 6 || ++num_scores >= NUM_SCORES)
			break;
	}
	if (!test_mode && !list_em) {
		if ((pw = (struct passwd *) getpwuid(getuid())) == NULL) {
			(void)fprintf(stderr,
				"getpwuid failed for uid %d.  Who are you?\n",
				(int)getuid());
			return (-1);
		}
		(void)strlcpy(thisscore.name, pw->pw_name, SCORE_NAME_LEN);
		(void)uname(&lname);
		(void)strlcpy(thisscore.host, lname.nodename,
		    sizeof(thisscore.host));

		cp = strrchr(filename, '/');
		if (cp == NULL) {
			(void)fprintf(stderr, "log: where's the '/' in %s?\n",
			    filename);
			return (-1);
		}
		cp++;
		(void)strlcpy(thisscore.game, cp, SCORE_GAME_LEN);

		thisscore.time = clck;
		thisscore.planes = safe_planes;
		thisscore.real_time = time(0) - start_time;

		for (i = 0; i < num_scores; i++) {
			if (strcmp(thisscore.name, score[i].name) == 0 &&
			    strcmp(thisscore.host, score[i].host) == 0 &&
			    strcmp(thisscore.game, score[i].game) == 0) {
				if (thisscore.time > score[i].time) {
					score[i].time = thisscore.time;
					score[i].planes = thisscore.planes;
					score[i].real_time =
					    thisscore.real_time;
					changed++;
				}
				found++;
				break;
			}
		}
		if (!found) {
			for (i = 0; i < num_scores; i++) {
				if (thisscore.time > score[i].time) {
					if (num_scores < NUM_SCORES)
						num_scores++;
					(void)memcpy(&score[num_scores - 1],
					    &score[i], sizeof (score[i]));
					(void)memcpy(&score[i], &thisscore,
					    sizeof (score[i]));
					changed++;
					break;
				}
			}
		}
		if (!found && !changed && num_scores < NUM_SCORES) {
			(void)memcpy(&score[num_scores], &thisscore,
			    sizeof (score[num_scores]));
			num_scores++;
			changed++;
		}

		if (changed) {
			if (found)
				(void)puts("You beat your previous score!");
			else
				(void)puts("You made the top players list!");
			qsort(score, (size_t)num_scores, sizeof (*score),
			    compar);
			rewind(score_fp);
			for (i = 0; i < num_scores; i++)
				(void)fprintf(score_fp, "%s %s %s %d %d %d\n",
				    score[i].name, score[i].host,
				    score[i].game, score[i].planes,
				    score[i].time, score[i].real_time);
			(void)fflush(score_fp);
			if (ferror(score_fp))
				warn("error writing %s", _PATH_SCORE);
			/* It is just possible that updating an entry could
			 * have reduced the length of the file, so we
			 * truncate it.  The seeks are required for stream/fd
			 * synchronisation by POSIX.1.  */
			offset = ftell(score_fp);
			(void)lseek(fileno(score_fp), (off_t)0, SEEK_SET);
			(void)ftruncate(fileno(score_fp), (off_t)offset);
			rewind(score_fp);
		} else {
			if (found)
				(void)puts(
				    "You didn't beat your previous score.");
			else
				(void)puts(
				    "You didn't make the top players list.");
		}
		(void)putchar('\n');
	}
#ifdef BSD
	(void)flock(fileno(score_fp), LOCK_UN);
#endif
#ifdef SYSV
	/* lock will evaporate upon close */
#endif
	(void)fclose(score_fp);
	(void)printf("%2s:  %-8s  %-8s  %-18s  %4s  %9s  %4s\n", "#", "name",
	    "host", "game", "time", "real time", "planes safe");
	(void)printf("-------------------------------------------------------");
	(void)printf("-------------------------\n");
	for (i = 0; i < num_scores; i++) {
		cp = strchr(score[i].host, '.');
		if (cp != NULL)
			*cp = '\0';
		(void)printf("%2d:  %-8s  %-8s  %-18s  %4d  %9s  %4d\n", i + 1,
		    score[i].name, score[i].host, score[i].game,
		    score[i].time, timestr(score[i].real_time),
		    score[i].planes);
	}
	(void)putchar('\n');
	return (0);
}

/* ARGSUSED */
void
log_score_quit(int dummy __unused)
{
	(void)log_score(0);
	exit(0);
}
